### ISSUE TITLE
Recursive definitions in mixins

### ISSUE THREAD

**Issue by SamuraiJack** [Microsoft/TypeScript#29872]
Its currently possible to create a circular references in classes just fine:
```ts
export class Class1 {
    another     : Class2     // compiles fine
}
export class Class2 {
    another     : Class1    // compiles fine
}
```
However, the same thing fails when using mixin-based classes:

```ts
export const SampleMixin1 = <T extends AnyConstructor<object>>(base : T) =>
class SampleMixin1 extends base {
    another             : SampleMixin2 // TS2502: 'another' is referenced directly or indirectly in its own type annotation
}
export type SampleMixin1 = Mixin<typeof SampleMixin1>


export const SampleMixin2 = <T extends AnyConstructor<object>>(base : T) =>
class SampleMixin2 extends base {
    another             : SampleMixin1 // TS2502: 'another' is referenced directly or indirectly in its own type annotation
}
export type SampleMixin2 = Mixin<typeof SampleMixin2>


// supporting declarations for mixin pattern
export type AnyFunction<A = any>      = (...input: any[]) => A
export type AnyConstructor<A = any>   = new (...input: any[]) => A
export type Mixin<T extends AnyFunction> = InstanceType<ReturnType<T>>
```

This makes things much more complicated, you need to introduce some dummy interfaces, etc, etc.

The workaround exists - to use different form of creating the type for the standalone mixin class - with interfaces:
```ts
export const SampleMixin3 = <T extends AnyConstructor<object>>(base : T) =>
class SampleMixin3 extends base {
    another             : SampleMixin4
}
export interface SampleMixin3 extends Mixin<typeof SampleMixin3> {}


export const SampleMixin4 = <T extends AnyConstructor<object>>(base : T) =>
class SampleMixin4 extends base {
    another             : SampleMixin3
}
export interface SampleMixin4 extends Mixin<typeof SampleMixin4> {}
```
But this notation seems to drive crazy the IDE (the language server under the hood?). It stop finding usages of properties, show many false type errors etc. Basically in such approach you are limited to old-good console `npx tsc` launch (which works well at least), but all the development aid facilities don't work.

I think mixin pattern should be supported first class. I'd even create some language construct for it.

**Comment by SamuraiJack** [comment]
Any hope for a proper solution of this issue?

**Comment by orta** [comment]
When using mixins via class expressions this doesn't [seem to apply](https://www.staging-typescript.org/play?strictPropertyInitialization=false#code/C4TwDgpgBAwg9gOwM7AE4FcDGw6qgXigQgHcoAKAOmoENUBzJALihoRAG0BdASgID4oAbwC+AbgBQEgGboE2AJaIoAZUw0ANhAA8AFQBCNJNAgAPYBAQATJLEQoM2XP3KHjLA0Yh8hEqFFQIYHRUBChMDSNbNU0FBHooMwtrWzdoX39-AHosqABZBVM42wBbGhAiOGAoKwgIumgwVAUANxoLLKaquosrKC7IVGAFCCQ-TJyoAAs4EggWiFQAGigQOHRwtih0YygAURUAJgAGE-7mtosoaRGNG3H-AH0kdS0CKABGSQeoY2AYrTkF6aCAsBDoEoAI0WPh+-mAUwUSEoz1e0EIwK0kkyUBEUhx9CCvzR5B4YIh0LwGRxASCITCCKRKMxEGxmTx-nEEjxMjkimUABk4PRCag9GlEuZLDY7Mg0FgcKgXGkPGlYf5AsFQuFIkhbEKRYtJckZRLfOMuTypPU9VAAOKodZgYTjJBNBQWZhQUCQODSfYKehTYD6D0qd0Wbjc6266IR9LjeiO9BgFg+iB+qAG+g0SFaB1O6MSTD2ap7QPB0P-ePvAEQcjh5oWHiSEtyrPCnN5iAFlPvbOLci9sAtiRAA):

```ts
type Constructor = new (...args: any[]) => {};

function Scale<TBase extends Constructor>(Base: TBase) {
  return class Scaling extends Base {
    _scale = 1;

    setScale(scale: number) {
      this._scale = scale;
    }

    get scale(): number {
      return this._scale;
    }
  };
}

function Logger<TBase extends Constructor>(Base: TBase) {
  return class Logger extends Base {

  };
}



class Group {
  sprites: typeof EightBitSprite[]
}

class Sprite {
  group: typeof LoggableGroup
}

const EightBitSprite = Scale(Sprite);
const LoggableGroup = Logger(Group);

```

**Comment by canonic-epicure** [comment]
@orta This is because you haven't created an "abstract" mixin type - a type that represents instance of any class that has consumed a mixin. You only work with specific classes. Abstract "Scalable" would look like:
```ts
// supporting declarations for mixin pattern
export type AnyFunction<A = any>      = (...input: any[]) => A
export type AnyConstructor<A = any>   = new (...input: any[]) => A
export type Mixin<T extends AnyFunction> = InstanceType<ReturnType<T>>

function Logger<TBase extends Constructor>(Base: TBase) {
  return class Logger extends Base {
 };
}

type LoggerT = Mixin<typeof Logger>  // "abstract" mixin type 
```

Actually, I just checked and the originally submitted code seems to work starting from 3.8.2!! (The version picker in playground is awesome). It does not work in 3.7.5 and prior. 

Hopefully this was an intentional improvement, properly covered with a test, not a random codebase fluctuation. Could you check may be?

**Comment by trusktr** [comment]
Yes please, I hope this is tested for. Robust moxins have been too brittle.

**Comment by canonic-epicure** [comment]
Hello again,

As I mentioned in the previous comment, the issue of this ticket seems to be resolved in 3.8.2 release. However, this ticket did not receive any updates at that point, so, I'm suspecting the "fix" might be just a side effect of some other change in the codebase.

Can someone from the dev team please check the originally reported issue and confirm, that this behavior is indeed now "officially" supported (and "pinned" with the test)?

**Comment by canonic-epicure** [comment]
No, actually, as I thought, the "fix" was just a random codebase fluctuation. First of all, if you enable `strictPropertyInitialization` it will fail again with the same message. W/o that option it works:
[Playground link](https://www.typescriptlang.org/play?noImplicitAny=false&declaration=false&ts=4.3.4#code/PTAEGcFcAdoewE4BcCWA7A5qAJgUwMYA2AhgsanGuKAGaKgC2KAHuqNOUrgmgFC7N4yUEgCe0XKACCaUQDFIafBTQAeKaAC8oYrIB8oQ0e0AKAHQX00SEgBcO2QG0AugEotBqf0GIkI8ZIyogDClOBICJDKiOpaDqIGhtpouADuoOaWaNZ28S7ump7eQn5iEqAAsizoqgAqoAJcaNjUQQpKKgbaAJJUSLr4uLUBqgBKuEiQPMMSdXp6vIsCJaD4YX4AysQM0IS4VaxoAIxxdQ3MTS3SsqF9kdEIqnAARgBWBEjzJs-E4JL2tQKCyIv2oWx2ewO6BOjVwzWoPz+oAA3rwjA44EgABbcdF49H2cG7fbVNAAJlAIFAtQ2ZIArAAGMn2ADkukxOIQLNAKGoCFwNG4cMG2BwKH5ykIolA9HQ2HFHylPLQPKQ1DgqRVZUkujQmPIKEovAAvsVfP5ykTIaSTtooWptXAaKArSTDkcFksfMI1n0XdtifaKdozrD4dcQut7kgYi93sovoj-tSgbwQeAwQHrYcKWGrkmUWijOzsbj8fjCVm3dDKWAafSmayS5zubzQPzBfylLhRfKJUglbLmgrJdK2Cg1TLNRadWg9f0VCazcJtf6IdXyXF7apHc7XUHPbwgA)

Secondly, it fails for 2nd level of mixins:
[Playground link](https://www.typescriptlang.org/play?noImplicitAny=false&strictPropertyInitialization=false&declaration=false&ts=4.2.3#code/PTAEGcFcAdoewE4BcCWA7A5qAJgUwMYA2AhgsanGuKAGaKgC2KAHuqNOUrgmgFC7N4yUEgCe0XKACCaUQDFIafBTQAeKaAC8oYrIB8oQ0e0AKAHQX00SEgBcO2QG0AugEotBqf0GIkI8ZIyogDClOBICJDKiOpaDqIGhtpouADuoOaWaNZ28S7ump7eQn5iEqAAsizoqgAqoAJcaNjUQQpKKgbaAJJUSLr4uLUBqgBKuEiQPMMSdXp6vMW+oPhhfgDKxAzQhLhVrGgAjHF1DcxNLdKyoX2R0QiqcABGAFYESPMmT8Tgkva1BQWRB+1E2212+3Qx0auGa1G+v1AAG9eEYHHAkAALbho3Fo+xgnZ7apoABMoBAoFq61JAFYAAyk+wAcl0GOxCGZoBQ1AQuBo3Fhg2wOBQfOUhFEoHo6GwYveku5aG5SGocFSyrKkl0aAx5BQlF4AF8lsItaBCRCScdtJC1Fq4DQLVsiXbDgtFgISis1s7wcSDuTtKcYXCriE1nckDFnm9lJ8EX8qYDeMDwKCXVbA2cLvCfpIUWi2VicXi8QTMwGoRSwNS6YyWcWOVyeaA+QK+UpcCK5eKkIqZc15RKpWwUKrpRr-OUdXqVMbTaUAn7XSSg5USaoHU7LVWyR7PT5hKs+ius+gAMwneqhy5BG7hKMxXduhP50D-FNpjP+u1X295oihZGE2pZlviZ57gALDWVI0gyTKgKyuolpy3K8vygpdj2w79lKg69gqo7KuOapTuas79POJpess5oviSV62pu26QX+B60cevoMQcMHBje5ywne1yRlE0YPDx6Ckm+iKfh4qYkOmbEkjBAGgImyKoiBKEcuBuIVr+jGwXWCGNjp3AthhHZCt2op9gOCBKoRI5KiqZGasulH6oaNFHku5SSWgfEbgcW4BI6ym8XoQA)

This is so frustrating. Probably a decent portion of the TS issues is related to the bad support of the mixin pattern. Why not finally implementing a proper solution/workaround? It does not have to be academically pure, it should just work. Can't believe its something impossible.

**Comment by trusktr** [comment]
Although the original example still does not work, the errors go away if written like [this](https://www.typescriptlang.org/play?noImplicitAny=false&declaration=false&ts=4.3.4#code/PTAEGcFcAdoewE4BcCWA7A5qAJgUwMYA2AhgsanGuKAGaKgC2KAHuqNOUrgmgFC7N4yUEgCe0XKACCaUQDFIafBTQAeKaAC8oYrIB8W0AAoAdGfTRISAFw7ZAbQC6ASi0Gp-QYiQjxkmaIAwpTgSAiQyoiqACqGcABGAFYESAA00qACXGjY1LqiTob5TukAykjkKPiGAN4AvgbaRmi4AO7GZibEtlKumgbRrgBkoOWV+LyeQj40isoolKPEDNCEuACyLOgAjDGZzNm50rLBVGERSIh6RvHE4Li2g6A1vKBvoAi4SJA8oER31FKy1WuG2+0O1Fu92er3ecN0cCQAAtuABCWwASTOunwuGiflUACUvj80PiJKoxBI4DQlis1ptWGgAEx6PSwt51XhcqbeWhzFR0kGM9DMvZZXA5agBU6hcKRBDXKEPUBPF5wz7fX7-cCA4FrZngyVHZUwuHwtCIlEIdGgLGhHF4gnErVkglU3A0oUMrZobZsjmgLl1IA):

```ts
// supporting declarations for mixin pattern
export type AnyFunction<A = any> = (...input: any[]) => A
export type AnyConstructor<T = object, A extends any[] = any[], Static = {}> = (new (...a: A) => T) & Static

export function SampleMixin1<T extends AnyConstructor>(base: T) {
    return class Sample1 extends base {
        another!: InstanceType<ReturnType<typeof SampleMixin2>>
    }
}

export function SampleMixin2<T extends AnyConstructor>(base: T) {
    return class Sample2 extends base {
        another!: InstanceType<ReturnType<typeof SampleMixin1>>
    }
}
```

However the result is very strange!!!! _**Sometimes**_ hovering on `another` will show proper types. Most of the time, it will just show `any`. There definitely something flaky happening under the hood.

Here is [something](https://www.typescriptlang.org/play?noImplicitAny=false&declaration=false&ts=4.3.4#code/PTAEGcFcAdoewE4BcCWA7A5qAJgUwMYA2AhgsanGuKAGaKgC2KAHuqNOUrgmgFC7N4yUEgCe0XKACCaUQDFIafBTQAeKaAC8oYrIB8W0AAoAdGfTRISAFw7ZAbQC6ASi0Gp-QYiQjxkmaIAwpTgSAiQyoiqACqGcABGAFYESAA00qACXGjY1LqiTob5TukAykjkKPiGAN4AvgbaRmi4AO7GZibEtlKumgbRrgBkoOWV+LyeQj40isoolKPEDNCEuACyLOgAjDGZzNm50rLBVGERSIh6RvHE4Li2g6A1vKBvoAi4SJA8oER31FKy1WuG2+0O1Fu92er3ecN0cCQAAtuABCWxAlZrABMsLedV4BKm3j+IR8mJBYO0FLWm1YaG2RgA8kkUs5icIxBIlljQYYubg4DQeZTJgJprQ5ioRbStmhsXssrgctQAqdQuFIghrlCHqAni84Z9vr9-uBAcCceDlUddTC4fC0IiUQh0TLQXjQASieKSfgye7sYYaRs5djmazlOzfZy-IH+X4hYGgA) I've been doing in practice that doesn't have any errors and seems to work (at least in my cases so far):

```ts
// supporting declarations for mixin pattern
export type AnyFunction<A = any> = (...input: any[]) => A
export type AnyConstructor<T = object, A extends any[] = any[], Static = {}> = (new (...a: A) => T) & Static

export function SampleMixin1<T extends AnyConstructor>(base: T) {
    return class Sample1 extends base {
        another!: Sample2
    }
}

export const Sample1 = SampleMixin1(Object)
export type Sample1 = typeof Sample1

export function SampleMixin2<T extends AnyConstructor>(base: T) {
    return class Sample2 extends base {
        another!: Sample1
    }
}

export const Sample2 = SampleMixin2(Object)
export type Sample2 = typeof Sample2
```

**Comment by canonic-epicure** [comment]
@trusktr Can you try making it working as in my 2nd example [here](https://www.typescriptlang.org/play?noImplicitAny=false&strictPropertyInitialization=false&declaration=false&ts=4.2.3#code/PTAEGcFcAdoewE4BcCWA7A5qAJgUwMYA2AhgsanGuKAGaKgC2KAHuqNOUrgmgFC7N4yUEgCe0XKACCaUQDFIafBTQAeKaAC8oYrIB8oQ0e0AKAHQX00SEgBcO2QG0AugEotBqf0GIkI8ZIyogDClOBICJDKiOpaDqIGhtpouADuoOaWaNZ28S7ump7eQn5iEqAAsizoqgAqoAJcaNjUQQpKKgbaAJJUSLr4uLUBqgBKuEiQPMMSdXp6vMW+oPhhfgDKxAzQhLhVrGgAjHF1DcxNLdKyoX2R0QiqcABGAFYESPMmT8Tgkva1BQWRB+1E2212+3Qx0auGa1G+v1AAG9eEYHHAkAALbho3Fo+xgnZ7apoABMoBAoFq61JAFYAAyk+wAcl0GOxCGZoBQ1AQuBo3Fhg2wOBQfOUhFEoHo6GwYveku5aG5SGocFSyrKkl0aAx5BQlF4AF8lsItaBCRCScdtJC1Fq4DQLVsiXbDgtFgISis1s7wcSDuTtKcYXCriE1nckDFnm9lJ8EX8qYDeMDwKCXVbA2cLvCfpIUWi2VicXi8QTMwGoRSwNS6YyWcWOVyeaA+QK+UpcCK5eKkIqZc15RKpWwUKrpRr-OUdXqVMbTaUAn7XSSg5USaoHU7LVWyR7PT5hKs+ius+gAMwneqhy5BG7hKMxXduhP50D-FNpjP+u1X295oihZGE2pZlviZ57gALDWVI0gyTKgKyuolpy3K8vygpdj2w79lKg69gqo7KuOapTuas79POJpess5oviSV62pu26QX+B60cevoMQcMHBje5ywne1yRlE0YPDx6Ckm+iKfh4qYkOmbEkjBAGgImyKoiBKEcuBuIVr+jGwXWCGNjp3AthhHZCt2op9gOCBKoRI5KiqZGasulH6oaNFHku5SSWgfEbgcW4BI6ym8XoQA)? Basically, create a `SampleMixin3` which extends `SampleMixin1` and `SampleMixin4`, which extends `SampleMixin2`. Then make a cycle between them
